autotest/pymod/gdaltest.py

```diff
@@ -2116,3 +2116,80 @@ def reopen(ds, update=False, open_options=None):
         allowed_drivers=[ds_drv.GetDescription()],
         open_options=open_options,
     )
+
+
+# VSIFile helper class
+
+
+class VSIFile:
+    def __init__(self, path, mode, encoding="utf-8"):
+        self._path = path
+        self._mode = mode
+
+        self._binary = "b" in mode
+        self._encoding = encoding
+
+        self._fp = gdal.VSIFOpenExL(self._path, self._mode, True)
+        if self._fp is None:
+            raise OSError(gdal.VSIGetLastErrorMsg())
+
+        self._closed = False
+        self._buffer = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        line = gdal.CPLReadLineL(self._fp)
+        if line is None:
+            raise StopIteration
+        if self._binary:
+            return line.encode()
+        return line
+
+    def close(self):
+        if self._closed:
+            return
+
+        self._closed = True
+        gdal.VSIFCloseL(self._fp)
+
+    def read(self, size=-1):
+        if size == -1:
+            pos = self.tell()
+            self.seek(0, 2)
+            size = self.tell()
+            self.seek(pos)
+
+        raw = gdal.VSIFReadL(1, size, self._fp)
+
+        if self._binary:
+            return bytes(raw)
+        else:
+            return raw.decode(self._encoding)
+
+    def write(self, x):
+
+        if self._binary:
+            assert type(x) in (bytes, bytearray, memoryview)
+        else:
+            assert type(x) is str
+            x = x.encode(self._encoding)
+
+        gdal.VSIFWriteL(x, 1, len(x), self._fp)
+
+    def seek(self, offset, whence=0):
+        gdal.VSIFSeekL(self._fp, offset, whence)
```

Comment:
we need to be careful here. Due to VSIFSeekL() taking only unsigned offsets,  negative seeking with SEEK_CUR must be emulated doing VSIFSeekL(fp, VSIFTellL(fp) + offset, SEEK_SET)

Reply:
Where should the emulation happen?

- `VSIFile.seek`
- `gdal.VSIFSeekL` (via `%pythonprepend%`)
- `gdal.VSIFSeekL` (via a new wrapper function in `cpl_i`. The signature here already uses signed offsets, so it seems reasonable that it should handle them correctly)

I'm still scratching my head a bit as to why the tests using negative seek work correctly as-is.

Reply:
> * `gdal.VSIFSeekL` (via a new wrapper function in `cpl_i`. The signature here already uses signed offsets, so it seems reasonable that it should handle them correctly)

sounds good here. By the way, I've just realized yesterday, that is is not needed to use a "wrapper_foo" + %rename tricks. You can directly create a inline function with the target name. cf https://github.com/OSGeo/gdal/pull/10050/commits/5bcbc5c05486bf02e0bdb212ca4dc2f6670c2188#diff-56ee4d91557b7ba46f850e46abbdc73e617c981c662a47a27d00eaef97136061R972

> I'm still scratching my head a bit as to why the tests using negative seek work correctly as-is.

I suppose that the negative offsets get turned into unsigned values within SWIG .cpp code, and then unsigned overflow on additions within the virtual file system implementations accidentally lead to the correct result. But we'd better not rely on that. In particular the CI sanitize target and our configuration of OSSFuzz errors out on unsigned integer overflows, as they are most of the time bugs.

Reply:
Added code to handle negative values in afe7ca8. I think the `wrapper_` and `%rename` are needed in this case because the function name exposed by SWIG is the same as the underlying function name.